src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/InfiniteScrollListBox.cs

```diff
@@ -1,12 +1,18 @@
 using System.Threading;
 using System.Windows.Automation.Peers;
 using System.Windows.Controls;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio;
 
 namespace NuGet.PackageManagement.UI
 { 
     internal class InfiniteScrollListBox : ListBox
     {
-        public readonly SemaphoreSlim ItemsLock = new SemaphoreSlim(1, 1);
+        public readonly ReentrantSemaphore ItemsLock =
+            ReentrantSemaphore.Create(
+                initialCount: 1,
+                joinableTaskContext: NuGetUIThreadHelper.JoinableTaskFactory.Context,
+                mode: ReentrantSemaphore.ReentrancyMode.Freeform);
```

Comment:
Considering that `Freeform` is the most permissive option and thus allows some user errors to go unnoticed, and considering that you only use this semaphore in exactly one place which doesn't seem to ever be recursive, have you considered using `NotAllowed` here?

Reply:
This semaphore is also used inside `InfiniteScrollList.xaml.cs` class. Basically whenever NuGet UI is loaded it starts to populate packages list which is where it's being used. So there might be few instance it could end up being recursive, which is why I thought `FreeForm` might be a better choice... Do you think it makes sense?

Reply:
I see, I didn't notice that.
If your recursion acts like a stack though, in that every time you enter the semaphore, your method doesn't return (or complete its async task) until it has exited the semaphore, then using the Stack enum value is better since it helps enforce that for you and then doesn't "leak" to any spin off work you might have done.

Reply:
hmm that makes sense I'll switch it to `Stack`